osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/services/SecretManagementService.java

```diff
@@ -35,147 +44,318 @@
+    public void storeNewKey(String deviceIdentification, SecurityKeyType keyType, byte[] key) {
+        Map<SecurityKeyType, byte[]> keysByType = new HashMap<>();
+        keysByType.put(keyType, key);
+        this.storeNewKeys(deviceIdentification, keysByType);
     }
 
-    @Override
-    public byte[] getMbusUserKey(String mbusDeviceIdentification) {
-        throw new NotImplementedException();
+    //@Override
+    public void storeNewKeys(String deviceIdentification, Map<SecurityKeyType, byte[]> keysByType) {
+        this.validateKeys(keysByType);
+        TypedSecrets typedSecrets = new TypedSecrets();
+        List<TypedSecret> typedSecretList = typedSecrets.getTypedSecret();
+        for (SecurityKeyType type : keysByType.keySet()) {
+            TypedSecret ts = new TypedSecret();
+            ts.setType(type.toSecretType());
+            ts.setSecret(this.encryptSoapSecret(keysByType.get(type),true));
+            typedSecretList.add(ts);
+        }
+        StoreSecretsRequest request = this.createStoreSecretsRequest(deviceIdentification, typedSecrets);
+        this.secretManagementClient.storeSecretsRequest(request);
     }
 
-    @Override
-    public byte[] getDlmsPassword(String deviceIdentification) {
-        throw new NotImplementedException();
+    private void validateKeys(Map<SecurityKeyType, byte[]> keysByType) {
+        long nrNulls = keysByType.values().stream().filter(key -> key == null).count();
+        if (nrNulls > 0) {
+            throw new IllegalArgumentException(
+                    String.format("Provided %s keys, %s of which were NULL", keysByType.size(), nrNulls));
+        }
     }
 
-    @Override
-    public DlmsDevice storeNewKey(DlmsDevice device, byte[] encryptedKey, SecurityKeyType keyType) {
-        throw new NotImplementedException();
+    private StoreSecretsRequest createStoreSecretsRequest(String deviceIdentification, TypedSecrets typedSecrets) {
+        StoreSecretsRequest request = new StoreSecretsRequest();
+        request.setDeviceId(deviceIdentification);
+        request.setTypedSecrets(typedSecrets);
+        return request;
     }
 
-    @Override
-    public DlmsDevice validateNewKey(DlmsDevice device, SecurityKeyType keyType) throws ProtocolAdapterException {
-        throw new NotImplementedException();
+    /**
+     * Updates the state of a new key from 'new' to 'active'
+     * <p>
+     * This method should be called to activate a new key stored with
+     * {@link #storeNewKeys(String, Map)} after it has
+     * been confirmed to be set on the device.
+     *
+     * @param deviceIdentification
+     *         DLMS device id
+     * @param keyType
+     *         type of key
+     *
+     * @throws ProtocolAdapterException
+     *         if no new key is stored with the given device
+     * @see #storeNewKeys(String, Map)
+     */
+    //@Override
+    public void activateNewKey(String deviceIdentification, SecurityKeyType keyType) throws ProtocolAdapterException {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'org.opensmartgridplatform.adapter.protocol.dlms.exceptions.ProtocolAdapterException', as it cannot be thrown from method's body. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)